python/ray/tune/utils/file_transfer.py

```diff
@@ -352,9 +352,22 @@ def _iter_remote(actor: ray.ActorID) -> Generator[bytes, None, None]:
 def _unpack_dir(stream: io.BytesIO, target_dir: str) -> None:
     """Unpack tarfile stream into target directory."""
     stream.seek(0)
-    with FileLock(f"{target_dir}.lock"):
-        with tarfile.open(fileobj=stream) as tar:
-            tar.extractall(target_dir)
+    target_dir = os.path.normpath(target_dir)
+    try:
+        # Timeout 0 means there will be only one attempt to acquire
+        # the file lock. If it cannot be aquired, a TimeoutError
+        # will be thrown.
+        with FileLock(f"{target_dir}.lock", timeout=0):
+            with tarfile.open(fileobj=stream) as tar:
+                tar.extractall(target_dir)
+    except TimeoutError:
+        # wait, but do not do anything
+        with FileLock(f"{target_dir}.lock"):
+            pass
+        # if the dir was locked due to being deleted,
+        # recreate
+        if not os.path.exists(target_dir):
+            _unpack_dir(stream, target_dir)
```

Comment:
I'm a bit worried about infinite recursion here. Can call again once, but set a flag `allow_retry=False` or so to make sure we don't run into the problem? (and otherwise raise an error)